dynamic_instrumentation/src/instrument_memory.rs

```diff
@@ -426,240 +479,175 @@ impl<'a, 'tcx: 'a> Visitor<'tcx> for CollectFunctionInstrumentationPoints<'a, 't
-                        let instr_operand = if let BorrowKind::Mut { .. } = bkind {
-                            InstrumentationOperand::Place(Operand::Copy(*p))
-                        } else {
-                            // Instrument immutable borrows by tracing the reference itself
-                            location.statement_index += 1;
-                            InstrumentationOperand::Reference(Operand::Copy(dest))
-                        };
-                        if has_outer_deref(p) {
-                            // this is a reborrow or field reference, i.e.
-                            // _2 = &(*_1) or _2 = &((*_1).0)
-                            let source = remove_outer_deref(*p, self.tcx);
-                            self.add_instrumentation_point(
-                                location,
-                                copy_fn,
-                                vec![instr_operand],
-                                false,
-                                false,
-                                EventMetadata {
-                                    source: Some(to_mir_place(&source)),
-                                    destination: Some(to_mir_place(&dest)),
-                                    transfer_kind: TransferKind::None,
-                                },
-                            );
-                        } else if !p.is_indirect() {
-                            self.add_instrumentation_point(
-                                location,
-                                addr_local_fn,
-                                vec![
-                                    instr_operand,
-                                    InstrumentationOperand::AddressUsize(make_const(
-                                        self.tcx,
-                                        p.local.as_u32(),
-                                    )),
-                                ],
-                                false,
-                                false,
-                                EventMetadata {
-                                    source: Some(to_mir_place(p)),
-                                    destination: Some(to_mir_place(&dest)),
-                                    transfer_kind: TransferKind::None,
-                                },
-                            );
-                        }
-                    }
-                    _ => (),
+            }
+            Rvalue::Cast(_, Operand::Copy(p) | Operand::Move(p), _)
+                if value_ty.is_integral() && !p.is_indirect() =>
+            {
+                if is_region_or_unsafe_ptr(place_ty(p)) {
+                    InstrumentationPoint::new(location, ptr_to_int_fn)
+                        .arg(InstrumentationArg::from_type(
+                            Operand::Copy(p.local.into()),
+                            &local_ty(p),
+                        ))
+                        .source(p)
+                        .add(self);
                 }
```

Comment:
This is technically a functional change, because if the first two clauses are true but the last is false, then the `Rvalue::Cast(_, op, _)` expression that's matched later on will get hit instead of it happening here. It doesn't make a difference in the snapshot currently, but I want to avoid making that change in this PR.